subprojects/core/src/main/java/org/gradle/api/internal/tasks/properties/DefaultPropertyWalker.java

```diff
@@ -104,12 +108,25 @@ private static void visitProperties(PropertyNode node, TypeMetadata typeMetadata
     private static void visitImplementation(PropertyNode node, PropertyVisitor visitor, PropertySpecFactory specFactory) {
         // The root bean (Task) implementation is currently tracked separately
         if (!node.isRoot()) {
-            DefaultTaskInputPropertySpec implementation = specFactory.createInputPropertySpec(node.getQualifiedPropertyName("class"), new ImplementationPropertyValue(node.getBean().getClass()));
+            DefaultTaskInputPropertySpec implementation = specFactory.createInputPropertySpec(node.getQualifiedPropertyName("class"), new ImplementationPropertyValue(getImplementationClass(node.getBean())));
             implementation.optional(false);
             visitor.visitInputProperty(implementation);
         }
     }
 
+    @VisibleForTesting
+    static Class<?> getImplementationClass(Object bean) {
```

Comment:
We should document what's happening here, i.e. explain why we need this for Groovy, and also that we don't need it for Java and Kotlin (but we might need something similar for a new language).

Reply:
I added an explaining comment.

Reply:
We are unpacking `Action` and `Callable` types differently. I think it would make sense to add tests for each, no?